layers/ipfix.go

```diff
@@ -0,0 +1,773 @@
+		return err
+	}
+
+	for len(data) > 0 {
+		h := &IPFIXSetHeader{}
+		if err := h.decodeFromBytes(&data); err != nil {
+			return err
+		}
+		switch h.ID {
+		case IPFIXSetIDTypeTemplate:
+			t := &IPFIXTemplateRecord{Header: *h}
+			if err := t.decodeFromBytes(&data); err == nil {
+				i.Templates[t.ID] = t
+			} else {
+				return err
+			}
+		case IPFIXSetIDTypeOptionTemplate:
+			ot := &IPFIXOptionTemplateRecord{Header: *h}
+			if err := ot.decodeFromBytes(&data); err == nil {
+				i.Templates[ot.ID] = ot
+			} else {
+				return err
+			}
+		default:
+			d := &IPFIXDataSet{Header: *h}
+			t := i.Templates[h.ID]
+			if err := d.decodeFromBytes(&data, t); err != nil {
+				return err
+			}
+			i.Data = append(i.Data, *d)
+		}
+	}
+	return nil
+}
+
+// IPFIXMessageHeader is a Message Header (RFC 7011 section 3.1)
+//
+// The format of the Message Header on the wire is:
+//
+//    0                   1                   2                   3
+//    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |       Version Number          |            Length             |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                           Export Time                         |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                       Sequence Number                         |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                    Observation Domain ID                      |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type IPFIXMessageHeader struct {
+	// Version of IPFIX to which this Message conforms.  The value of
+	// this field is 0x000a for the current version, incrementing by one
+	// the version used in the NetFlow services export version 9
+	Version uint16
+	// Total length of the IPFIX Message, measured in octets, including
+	// Message Header and Set(s).
+	Length uint16
+	// Time at which the IPFIX Message Header leaves the Exporter,
+	// expressed in seconds since the UNIX epoch of 1 January 1970 at
+	// 00:00 UTC, encoded as an unsigned 32-bit integer.
+	ExportTime uint32
+	// Incremental sequence counter modulo 2^32 of all IPFIX Data Records
+	// sent in the current stream from the current Observation Domain by
+	// the Exporting Process.
+	SequenceNumber uint32
+	// A 32-bit identifier of the Observation Domain that is locally
+	// unique to the Exporting Process.  The Exporting Process uses the
+	// Observation Domain ID to uniquely identify to the Collecting
+	// Process the Observation Domain that metered the Flows.
+	ObservationDomainID uint32
+}
+
+// Len returns the data length in octet required by IPFIX header message
+func (mh IPFIXMessageHeader) Len() int {
+	return 2 + 2 + 4 + 4 + 4
+}
+
+func (mh *IPFIXMessageHeader) decodeFromBytes(data *[]byte) error {
```

Comment:
in go a slice is actually a pointer to an array, so your data here doesnt need to be *[]byte, but can just be []byte as that is actually a pointer to the struct anyway(under the hood).  unless you need to pass the *[]byte for some reason, i would change it to just be []byte so that things are cleaner and less chance for mistake later in the constant dereferencing.  same thing for all other instances of this below

Reply:
When we pass a slice to a function as an argument the values of the slice are passed by reference, but all the metadata describing the slice itself are just copies. And here any `decodeFromBytes` method will also change the slice metadata, they are reducing the slice from the start of the data they decoded.

``` go
func main() {
	slice := []string{"a", "a"}

	func(slice []string) {
		slice = slice[1:]
		fmt.Println(slice)
	}(slice)
	fmt.Println(slice)
}
```
will print:
```
[a]
[a a]
```

Reply:
that is correct. it didnt look like you cared about updating the metadata in the slice, hence my comment.  All good though.